File: day02/2.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <ranges>
#include <string_view>
#include <unordered_map>

using namespace std;

int main(int argc, char **argv) {
    fstream file(argv[1]);
    string line;
    int product_sum = 0;
    while (getline(file, line)) {
        string_view sv{line};
        string_view sets = sv.substr(sv.find_first_of(':') + 1);
        unordered_map<string_view, int> colors = {
            {"red", 0},
            {"green", 0},
            {"blue", 0},
        };
        for (const auto &set : ranges::split_view(sets, ';')) {
            for (const auto &c : ranges::split_view(set, ',')) {
                string_view cubes{c.begin(), c.end()};
                int count =
                    stoi(cubes.substr(1, cubes.find_last_of(' ')).data());
                string_view color = cubes.substr(cubes.find_last_of(' ') + 1);
                if (colors[color] < count) {
                    colors[color] = count;
                }
            }
        }
        product_sum += ranges::fold_left(
            colors, 1,
            [](int product, const auto &p) { return product * p.second; });
    }
    cout << product_sum << "\n";
    return 0;
}

File: day02/2.py
import sys
import re
import math

file = open(sys.argv[1])
limits = {"red": 12, "green": 13, "blue": 14}
product_sum = 0
for line in file:
    cubes = re.split(";|,", line.split(":")[1])
    minis = {"red": 0, "green": 0, "blue": 0}
    for cube in cubes:
        count, color = cube.strip().split(" ")
        count = int(count)
        if minis[color] < count:
            minis[color] = count
    product_sum += math.prod(minis.values())
print(product_sum)
